### PR TITLE
storage,roachpb: expose pebble.InternalIteratorStats for scans

### PR DESCRIPTION
**Pull request by sumeerbhola**
The current IteratorStats only include information from the top
of the tree (pebble.Iterator). The new InternalIteratorStats
includes block bytes loaded (and cached), and stats for points
iterated over when merging across levels in the log-structured
merge tree. The latter includes points that were ignored because
of range tombstones, but could not be cheaply skipped.

These stats can be used to make inferences about root causes
of slow scans.

This change exposes these in roachpb.ScanStats and the trace
span recorded when doing a mvcc get or scan. There are todos for
plumbing all or part of it through the higher layers via
roachpb.ScanStats and execinfrapb.KVStats.

Informs #59069
Informs https://github.com/cockroachdb/pebble/issues/1342

Release justification: low risk, and potentially high benefit
observability increase for existing functionality.

Release note: None

### CODE REVIEW

**Review by sumeerbhola — COMMENTED** [review]
This is a trivial PR, that exposes additional stats for scans.
@yuzefovich @azhng : I am including you here because you reviewed https://github.com/cockroachdb/cockroach/pull/64503 by @jordanlewis, that initially added plumbing for stats.
@andreimatei : including you since you have expressed interest in the past, and because this is adding a slightly bigger tracing payload.
I would like to get this in for v22.1. One question is whether we have enough tracing for customer deployments that we will be able to see these stats and make inferences, or do we think we need to do the additional upstream plumbing that is currently listed as todos.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77512#-:-MxfkvUu-lhBF_l3kAfA:b-bv6iky)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @Azhng, and @yuzefovich)

**Comment by yuzefovich** [comment]
I'll defer to Archer on the observability side of things (my only comment on #64503 was [this](https://github.com/cockroachdb/cockroach/pull/64503#pullrequestreview-752746274) which I made after that PR was merged).

**Review by nicktrav — APPROVED** [review]
pkg/storage :lgtm: 

> Reviewed 6 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77512#-:-Mxg0PSN5XLGYGElREEW:bjfsufb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @Azhng, and @yuzefovich)

**Review by sumeerbhola — COMMENTED** [review]
> I'll defer to Archer on the observability side of things (my only comment on #64503 was this which I made after that PR was merged).

I don't have a good answer to that:
- The member variable of type `execinfra.ScanStats` in `joinReader` looks unnecessary. It is read into and then read from in the same method. The same is true of `invertedJoiner`.
- The implementations of `execStatsForTrace` in both `joinReader` and `invertedJoiner` look similar, which relates to your question.
- These implementations do hook into the same `ProcessorBaseNoHelper.ExecStatsForTrace`, so at least there is generalization on when they are called.
- The changes in my PR don't propagate to `execinfra.ScanStats` and `execinfrapb.KVStats`. Even if we add those, I don't think it will change any of the integration in these joiners.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77512#-:-Mxj2QSqE9knL_vPzULd:b-ewdcir)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @Azhng, and @yuzefovich)

**Review by Azhng — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77512#-:-MxjYJctBvmq-jzg1OwO:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @andreimatei, @sumeerbhola, and @yuzefovich)

___
*[pkg/sql/execinfra/stats.go, line 60 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/77512#-MxjY9a56YmwqIFcaTDt:-MxjY9a56YmwqIFcaTDu:b-41mrnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f22c96d5efcf0439f962eb7c31d3c535e4b286da/pkg/sql/execinfra/stats.go#L60)):*
> ```Go
> // of each stat.
> // TODO(sql-observability): include other fields that are in roachpb.ScanStats,
> // here and in execinfrapb.KVStats.
> ```

How urgent does this need to be plumbed through?

**Comment by sumeerbhola** [comment]
> How urgent does this need to be plumbed through?

It depends on how useful it is to only have the current trace statements. Do you know the answer?

**Comment by sumeerbhola** [comment]
TFTRs!

**Comment by sumeerbhola** [comment]
bors r=azhng,yuzefovich,nicktrav

**Comment by Azhng** [comment]
> It depends on how useful it is to only have the current trace statements. Do you know the answer?

Hmm I don't have a definite answer :thinking: . I filed #77580 so we can track the plumbing work.

**Pull request merged**
